src/Compilers/Core/Portable/Compilation/DeterministicKey.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class SyntaxTreeKey
+    {
+        public abstract string FilePath { get; }
+        public abstract ParseOptions Options { get; }
+
+        protected SyntaxTreeKey()
+        {
+        }
+
+        public abstract SourceText GetText(CancellationToken cancellationToken = default);
+
+        public static SyntaxTreeKey Create(SyntaxTree tree)
+            => new DefaultSyntaxTreeKey(tree);
+
+        private sealed class DefaultSyntaxTreeKey : SyntaxTreeKey
+        {
+            private readonly SyntaxTree _tree;
+
+            public DefaultSyntaxTreeKey(SyntaxTree tree)
+            {
+                _tree = tree;
+            }
+
+            public override string FilePath
+                => _tree.FilePath;
+
+            public override ParseOptions Options
+                => _tree.Options;
+
+            public override SourceText GetText(CancellationToken cancellationToken = default)
+                => _tree.GetText(cancellationToken);
+        }
+    }
+
+    internal static class DeterministicKey
```

Comment:
this would be public.  I personally prefer this as the 'sibling components entrypoint' where 'Compilation' can be the instance-entrypoint for people who have that.